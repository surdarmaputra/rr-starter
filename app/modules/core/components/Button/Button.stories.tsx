import type { Meta, StoryObj } from '@storybook/react';

import { Button, type ButtonProps } from './Button';

type Story = StoryObj<typeof Button>;

const meta: Meta = {
  component: Button,
  title: 'Core/Components/Button',
};

export default meta;

export const Default: Story = {
  args: {
    children: 'Button',
  },
  render: (args: ButtonProps) => <Button {...args} />,
};
