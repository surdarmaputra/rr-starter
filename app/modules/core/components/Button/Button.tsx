import { Slot } from '@radix-ui/react-slot';
import { cva, type VariantProps } from 'class-variance-authority';
import * as React from 'react';

import { cn } from '~/modules/core/libs/shadcn/utils';

import type { TestableComponentProps } from '../../types';

const buttonVariants = cva(
  `
    inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-full text-sm font-semibold
    ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2
    focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:pointer-events-none
    disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0
    dark:ring-offset-slate-950 dark:focus-visible:ring-slate-300
  `,
  {
    variants: {
      variant: {
        primary: 'bg-primary text-white hover:bg-primary-hover',
        'primary-outlined':
          'border border-primary text-primary hover:bg-primary-hover hover:text-white',
        'primary-ghost': 'text-primary hover:bg-primary-hover hover:text-white',
        secondary: 'bg-secondary text-white hover:bg-secondary-hover',
        'secondary-outlined':
          'border border-secondary text-secondary hover:bg-secondary-hover hover:text-white',
        'secondary-ghost':
          'text-secondary hover:bg-secondary-hover hover:text-white',
        success: 'bg-success text-white hover:bg-success-hover',
        'success-outlined':
          'border border-success text-success hover:bg-success-hover hover:text-white',
        'success-ghost': 'text-success hover:bg-success-hover hover:text-white',
        warning: 'bg-warning text-white hover:bg-warning-hover',
        'warning-outlined':
          'border border-warning text-warning hover:bg-warning-hover hover:text-white',
        'warning-ghost': 'text-warning hover:bg-warning-hover hover:text-white',
        danger: 'bg-danger text-white hover:bg-danger-hover',
        'danger-outlined':
          'border border-danger text-danger hover:bg-danger-hover hover:text-white',
        'danger-ghost': 'text-danger hover:bg-danger-hover hover:text-white',
        link: 'text-textTitle underline-offset-4 hover:underline',
      },
      size: {
        default: 'h-10 px-4 py-2',
        sm: 'h-9 px-3',
        lg: 'h-11 px-8',
        icon: 'h-10 w-10',
      },
    },
    defaultVariants: {
      variant: 'primary',
      size: 'default',
    },
  },
);

export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement>,
    VariantProps<typeof buttonVariants>,
    TestableComponentProps {
  asChild?: boolean;
}

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  ({ className, variant, size, asChild = false, ...props }, ref) => {
    const Comp = asChild ? Slot : 'button';
    return (
      <Comp
        className={cn(buttonVariants({ variant, size, className }))}
        ref={ref}
        {...props}
      />
    );
  },
);

Button.displayName = 'Button';

export { Button, buttonVariants };
