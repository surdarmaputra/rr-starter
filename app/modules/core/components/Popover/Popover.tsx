import * as PopoverPrimitive from '@radix-ui/react-popover';
import * as React from 'react';

import { cn } from '~/libs/shadcn/utils';

export type PopoverProps = React.ComponentPropsWithoutRef<
  typeof PopoverPrimitive.Root
>;

const Popover = PopoverPrimitive.Root;

const PopoverTrigger = PopoverPrimitive.Trigger;

export type PopoverContentProps = React.ComponentPropsWithoutRef<
  typeof PopoverPrimitive.Content
>;

const PopoverContent = React.forwardRef<
  React.ComponentRef<typeof PopoverPrimitive.Content>,
  PopoverContentProps
>(
  (
    { children, className, align = 'center', sideOffset = 5, ...props },
    ref,
  ) => (
    <PopoverPrimitive.Portal>
      <PopoverPrimitive.Content
        align={align}
        className={cn(
          'z-50 w-72 rounded-lg bg-white p-4 text-letter-body shadow-md outline-none dark:bg-accent',
          'data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0',
          'data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95',
          'data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2',
          'data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2',
          className,
        )}
        ref={ref}
        sideOffset={sideOffset}
        {...props}
      >
        {children}
        <PopoverPrimitive.Arrow className="fill-white dark:fill-accent" />
      </PopoverPrimitive.Content>
    </PopoverPrimitive.Portal>
  ),
);
PopoverContent.displayName = PopoverPrimitive.Content.displayName;

export { Popover, PopoverContent, PopoverTrigger };
