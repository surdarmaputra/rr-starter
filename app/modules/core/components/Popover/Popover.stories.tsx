import type { Meta, StoryObj } from '@storybook/react';

import { Button } from '../Button/Button';
import {
  Popover,
  PopoverContent,
  type PopoverProps,
  PopoverTrigger,
} from './Popover';

type Story = StoryObj<typeof Popover>;

const meta: Meta = {
  component: Popover,
  title: 'Core/Components/Popover',
};

export default meta;

export const Default: Story = {
  render: (args: PopoverProps) => (
    <Popover {...args}>
      <PopoverTrigger asChild>
        <Button>Click me</Button>
      </PopoverTrigger>
      <PopoverContent data-testid="PopoverContent-testid">
        Popover Content
      </PopoverContent>
    </Popover>
  ),
};
