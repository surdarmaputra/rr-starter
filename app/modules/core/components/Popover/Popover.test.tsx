import { fireEvent, render, screen } from '@testing-library/react';
import { describe, expect, test } from 'vitest';

import { Button } from '../Button/Button';
import { Popover, PopoverContent, PopoverTrigger } from './Popover';

describe('Popover', () => {
  function renderComponent() {
    return render(
      <Popover>
        <PopoverTrigger asChild>
          <Button>Action</Button>
        </PopoverTrigger>
        <PopoverContent data-testid="PopoverContent-testid">
          Popover Content
        </PopoverContent>
      </Popover>,
    );
  }

  test('renders content when trigger is clicked', async () => {
    renderComponent();
    expect(
      screen.queryByTestId('PopoverContent-testid'),
    ).not.toBeInTheDocument();
    expect(screen.queryByText(/Popover Content/)).not.toBeInTheDocument();

    fireEvent.click(screen.getByRole('button', { name: /Action/ }));
    await screen.findByTestId('PopoverContent-testid');
    screen.getByText(/Popover Content/);
  });

  test('renders snapshot', async () => {
    renderComponent();
    fireEvent.click(screen.getByRole('button', { name: /Action/ }));
    const popover = await screen.findByTestId('PopoverContent-testid');
    expect(popover).toMatchSnapshot();
  });
});
