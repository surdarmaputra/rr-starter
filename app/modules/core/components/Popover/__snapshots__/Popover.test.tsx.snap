// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Popover > renders snapshot 1`] = `
<div
  class="z-50 w-72 rounded-lg bg-white p-4 text-letter-body shadow-md outline-none dark:bg-accent data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2"
  data-align="center"
  data-side="bottom"
  data-state="open"
  data-testid="PopoverContent-testid"
  id="radix-:r1:"
  role="dialog"
  style="--radix-popover-content-transform-origin: var(--radix-popper-transform-origin); --radix-popover-content-available-width: var(--radix-popper-available-width); --radix-popover-content-available-height: var(--radix-popper-available-height); --radix-popover-trigger-width: var(--radix-popper-anchor-width); --radix-popover-trigger-height: var(--radix-popper-anchor-height);"
  tabindex="-1"
>
  Popover Content
  <span
    style="position: absolute; top: 0px; transform-origin: center 0; transform: rotate(180deg); left: 0px;"
  >
    <svg
      class="fill-white dark:fill-accent"
      height="5"
      preserveAspectRatio="none"
      style="display: block;"
      viewBox="0 0 30 10"
      width="10"
    >
      <polygon
        points="0,0 30,0 15,10"
      />
    </svg>
  </span>
</div>
`;
